feat: add dynamicTag for RocketMQ

wyc194 opened the pull request:
提交了dynamicTag的支持，使用dynamicTopic相同的匹配分隔逻辑。遗憾的是，对于目前支持的所有MQ，感觉仅RocketMQ对tag的支持是更有意义的，因此该提交仅处理了CanalRocketMQProducer的支持。这样的实现看起来有些异味，只是目前存在这样的需求，对RocketMQ同个topic下的消息标记tag，消费端按需过滤不同的tag。另外在issues中也看到一些同样的对dynamicTag的需求，如：[5375](https://github.com/alibaba/canal/issues/5375)，[4468](https://github.com/alibaba/canal/issues/4468)，[3409](https://github.com/alibaba/canal/issues/3409)，[1793](https://github.com/alibaba/canal/issues/1793) 等。
请评估下是否可以支持该功能，谢谢。

wyc194 committed:
feat: add dynamicTag for RocketMQ

wenshao commented on connector/rocketmq-connector/src/main/java/com/alibaba/otter/canal/connector/rocketmq/producer/CanalRocketMQProducer.java:
```diff
@@ -182,8 +204,37 @@ public void send(MQDestination destination, com.alibaba.otter.canal.protocol.Mes
+                                  String dynamicTagConfigs) {
+        // 动态tag, 直接使用[动态topic]相同的分隔逻辑
+        Map<String, com.alibaba.otter.canal.protocol.Message> messageMap = MQMessageUtils
+            .messageTopics(message, null, dynamicTagConfigs);
```
**[Critical]** `sendByDynamicTag` passes `null` as the `defaultTopic` argument to `MQMessageUtils.messageTopics()`. When any entry does not match a configured `dynamicTag` pattern (or has an empty schema/table name, e.g. DDL/heartbeat events), `messageTopics` falls through to `put2MapMessage(messages, id, null, entry)`, inserting a `null` key into the returned `HashMap`. The next line — `entry.getKey().replace('.', '_')` — then throws `NullPointerException`, crashing the entire send batch and triggering `callback.rollback()`.

**Failure scenario:** Configure `canal.mq.dynamicTag=mytest1.user` (no catch-all `.*\\..*` pattern). A binlog event for `mytest2.orders` arrives → no pattern match → `null` map key → NPE → batch rollback → infinite retry loop permanently blocking the canal instance. The existing `dynamicTopic` path avoids this by passing `destination.getTopic()` as the default.

```suggestion
        Map<String, com.alibaba.otter.canal.protocol.Message> messageMap = MQMessageUtils
            .messageTopics(message, ((RocketMQProducerConfig) this.mqProperties).getTag(), dynamicTagConfigs);
```

<details><summary>中文说明</summary>

`sendByDynamicTag` 向 `MQMessageUtils.messageTopics()` 传递 `null` 作为 `defaultTopic` 参数。当任何条目不匹配已配置的 `dynamicTag` 模式（或 schema/table 名为空，例如 DDL/心跳事件）时，`messageTopics` 会走到 `put2MapMessage(messages, id, null, entry)` 的兜底逻辑，在返回的 `HashMap` 中插入一个 `null` 键。紧接着的 `entry.getKey().replace('.', '_')` 会抛出 `NullPointerException`，导致整个发送批次失败并触发 `callback.rollback()`。

**故障场景：** 配置 `canal.mq.dynamicTag=mytest1.user`（未配置通配 `.*\\..*`）。当 `mytest2.orders` 的 binlog 事件到达时 → 无模式匹配 → `null` 键 → NPE → 批次回滚 → 无限重试循环，永久阻塞 canal 实例。现有的 `dynamicTopic` 路径通过传递 `destination.getTopic()` 作为默认值避免了此问题。

建议将 `null` 替换为已配置的静态 tag（`((RocketMQProducerConfig) this.mqProperties).getTag()`），使未匹配的条目回退到默认 tag。

</details>

_— qwen3.8-max-preview via Qwen Code /review_

wenshao commented on connector/rocketmq-connector/src/main/java/com/alibaba/otter/canal/connector/rocketmq/producer/CanalRocketMQProducer.java:
```diff
@@ -159,18 +159,40 @@ public void send(MQDestination destination, com.alibaba.otter.canal.protocol.Mes
-                            throw new RuntimeException(e);
-                        }
-                    });
+                    if (!StringUtils.isEmpty(destination.getDynamicTag())) {
+                        // 按动态tag发送
+                        sendByDynamicTag(template, destination, messageSub, topicName, destination.getDynamicTag());
```
**[Suggestion]** No test exercises the new four-way branching in `send()`: dynamicTopic+dynamicTag, dynamicTopic-only, dynamicTag-only, and neither. The entire `connector/rocketmq-connector` module has zero test files. — Concrete cost: a regression in branch ordering or a missing `template.waitForResult()` on one path would silently route messages to the wrong topic/tag combination, with no test to detect it.

Consider adding unit tests for `CanalRocketMQProducer.send()` covering all four branches, mocking `MQMessageUtils.messageTopics` and the underlying RocketMQ producer.

<details><summary>中文说明</summary>

没有测试覆盖 `send()` 中新增的四路分支逻辑：dynamicTopic+dynamicTag、仅 dynamicTopic、仅 dynamicTag、两者均无。整个 `connector/rocketmq-connector` 模块没有任何测试文件。具体风险：分支顺序的回归或某条路径上缺少 `template.waitForResult()` 会静默地将消息路由到错误的 topic/tag 组合，且没有测试能够检测到。

建议为 `CanalRocketMQProducer.send()` 添加单元测试，覆盖所有四个分支。

</details>

_— qwen3.8-max-preview via Qwen Code /review_

wyc194 committed:
Update connector/rocketmq-connector/src/main/java/com/alibaba/otter/canal/connector/rocketmq/producer/CanalRocketMQProducer.java

sendByDynamicTag 向 MQMessageUtils.messageTopics() 传递 null 替换为传递已配置的静态 tag（((RocketMQProducerConfig) this.mqProperties).getTag()），使未匹配的条目回退到默认 tag

Co-authored-by: Shaojin Wen <shaojin.wensj@alibaba-inc.com>
